### ISSUE TITLE
File descriptor leak

### ISSUE THREAD

**Issue by peterellisjones** [bosh-prometheus/bosh_exporter#14]
Hi bosh_exporter devs,

We're seeing the bosh_exporter process fail due to leaking file descriptors. For example here is the output of `lsof -c bosh_exporter`

```
...
bosh_expo 16454 root    6u  sock     0,8      0t0 1620523 can't identify protocol
bosh_expo 16454 root    7u  sock     0,8      0t0 1620524 can't identify protocol
bosh_expo 16454 root    8u  sock     0,8      0t0 1620527 can't identify protocol
bosh_expo 16454 root    9u  sock     0,8      0t0 1620633 can't identify protocol
bosh_expo 16454 root   10u  sock     0,8      0t0 1620756 can't identify protocol
bosh_expo 16454 root   11u  sock     0,8      0t0 1620864 can't identify protocol
bosh_expo 16454 root   12u  sock     0,8      0t0 1620972 can't identify protocol
bosh_expo 16454 root   13u  sock     0,8      0t0 1621079 can't identify protocol
bosh_expo 16454 root   14u  sock     0,8      0t0 1621188 can't identify protocol
bosh_expo 16454 root   15u  sock     0,8      0t0 1621386 can't identify protocol
bosh_expo 16454 root   16u  sock     0,8      0t0 1621497 can't identify protocol
...
```
This causes the scrape to fail, and we see the following in `/var/vcap/sys/log/bosh_exporter/bosh_exporter.stderr.log` 

```
...
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 5ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 10ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 20ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 40ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 80ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 160ms
2017/11/15 11:34:05 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 320ms
2017/11/15 11:34:06 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 640ms
...
```

https://stackoverflow.com/a/13789866 suggests this might be due to half-open TCP/IP connections which I'm guessing may be being leaked by the bosh-cli client.

Our stop-gap solution is to restart the monit process whenever it fails. 

We're running prometheus bosh-release v19.0.1 on AWS using the ubuntu-trusty stemcell 3445.16

cheers,

Pete

**Comment by frodenas** [comment]
From @marcus-sds at https://github.com/cloudfoundry-community/prometheus-boshrelease/issues/134:

Hi, still we have file descriptor issues with bosh_exporter. The version is prometheus/18.6.2.
1. lsof result exceeds limit : 1027
2. monit restart bosh_exporter
3. lsof check
prometheus/d32d13ce-5802-461c-9a1c-f442e3716b0b:/var/vcap/jobs/bosh_exporter/packages# lsof -p 22086|wc -l
9
prometheus/d32d13ce-5802-461c-9a1c-f442e3716b0b:/var/vcap/jobs/bosh_exporter/packages# lsof -p 22086|wc -l
24
prometheus/d32d13ce-5802-461c-9a1c-f442e3716b0b:/var/vcap/jobs/bosh_exporter/packages# lsof -p 22086|wc -l
19
prometheus/d32d13ce-5802-461c-9a1c-f442e3716b0b:/var/vcap/jobs/bosh_exporter/packages# lsof -p 22086|wc -l
17
prometheus/d32d13ce-5802-461c-9a1c-f442e3716b0b:/var/vcap/jobs/bosh_exporter/packages# lsof -p 22086|wc -l
25
4. netstat result
netstat -an|grep 9190|wc -l
1

========== error log
2017/10/19 13:02:52 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 320ms
2017/10/19 13:02:52 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 640ms
2017/10/19 13:02:53 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 1s
2017/10/19 13:02:54 http: Accept error: accept tcp 0.0.0.0:9190: accept4: too many open files; retrying in 1s
time="2017-10-19T13:02:54Z" level=error msg="Error while reading deployments: Finding deployments: Performing request GET 'https://10.10.0.6:25555/deployments': Performing GET request: Retry: Get https://10.10.0.6:25555/deployments: dial tcp 10.10.0.6:25555: socket: too many open files" source="bosh_collector.go:167"

**Comment by frodenas** [comment]
From @marcus-sds at https://github.com/cloudfoundry-community/prometheus-boshrelease/issues/134#issuecomment-345488333:

With workaround, I added cron release opsfile like below.

- type: replace
  path: /instance_groups/name=prometheus/jobs/-
  value:
    name: cron
    release: cron
    properties:
      cron:
        variables:
          PATH: /var/vcap/package/mypackage/bin:/bin:/sbin:/usr/sbin:/usr/bin:/usr/local/bin
        entries:
        - command: /var/vcap/bosh/bin/monit restart bosh_exporter
          minute: '59'
          hour: '15'
          day: '*'
          month: '*'
          wday: '*'
          user: root
        - command: /var/vcap/bosh/bin/monit restart cf_exporter
          minute: '59'
          hour: '15'
          day: '*'
          month: '*'
          wday: '*'
          user: root

- type: replace
  path: /releases/-
  value:
    name: cron
    version: "latest"

**Comment by frodenas** [comment]
@peterellisjones @marcus-sds before the `too many open files` error message, are there any other error messages? I've been running a `bosh_exporter` for a while, and have not experienced this, so I'm wondering if this is a special case where the process is failing to talk to the director and then it leaks connections.

**Comment by frodenas** [comment]
Another question @peterellisjones @marcus-sds: are you using `username` and `password` or UAA `client_id` and `client_secret`?

**Comment by takeyourhatoff** [comment]
@frodenas Hi, I will be taking ownership of this issue from @peterellisjones.
We are using client_id and client_secret for UAA.

**Comment by peterellisjones** [comment]
> @peterellisjones @marcus-sds before the too many open files error message, are there any other error messages?

Hi @frodenas — we didn't see any other errors. We've added your cron-release work-around and it's working fine so thanks!

**Comment by aeijdenberg** [comment]
This was affecting us as well, so I spent some time looking into. I'm pretty sure I've identified at least one source of file descriptor leaks that affect those using UAA, and I've sent this PR to the `bosh-cli` team: <https://github.com/cloudfoundry/bosh-cli/pull/383>

**Comment by aeijdenberg** [comment]
Fix has landed for at least one source of file descriptor leaks in `bosh-cli`. I guess we'll need to either wait for their next release to be cut (presumably 2.0.49) or update `dep` to pin to latest `develop`?

**Comment by frodenas** [comment]
I prefer not to use `develop`, let's wait until a new release is cut.

**Comment by frodenas** [comment]
This has been fixed at [v3.0.1](https://github.com/bosh-prometheus/bosh_exporter/releases/tag/v3.0.1)
